README.md
# Sensu StatsD Extension

A StatsD implementation for the Sensu client. This check extension
creates a StatsD TCP & UDP listener, receives StatsD metrics, parses
them, and flushes them to the Graphite plaintext format for Sensu to
send to Graphite or another TSDB.

This StatsD implementation attempts to adhere to [Etsy's metric type
specifications](https://github.com/etsy/statsd/blob/master/docs/metric_types.md).

[![Build Status](https://travis-ci.org/sensu-extensions/sensu-extensions-statsd.svg?branch=master)](https://travis-ci.org/sensu-extensions/sensu-extensions-statsd)

## Installation

This extension requires Sensu version >= 0.26.

On a Sensu client machine.

```
sensu-install -e statsd:1.0.0
```

Edit `/etc/sensu/conf.d/extensions.json` to load it.

``` json
{
  "extensions": {
    "statsd": {
      "version": "1.0.0"
    }
  }
}
```

Restart the Sensu client.

``` shell
sudo service sensu-client restart
```

## Configuration

Edit `/etc/sensu/conf.d/statsd.json` to change its configuration.

``` json
{
  "statsd": {
    "bind": "0.0.0.0"
  }
}
```

The following defaults make the integration behave like Etsy's StatsD
implementation.

|attribute|type|default|description|
|----|----|----|---|
|bind|string|"127.0.0.1"|IP to bind the StatsD sockets to|
|port|integer|8125|Port to bind the StatsD sockets to|
|flush_interval|integer|10|The StatsD flush interval|
|send_interval|integer|30|How often Graphite metrics are sent to Sensu|
|percentile|integer|90|The percentile to calculate for StatsD metrics|
|add_client_prefix|boolean|true|If the Sensu client name should prefix the Graphite metric path|
|path_prefix|string|"statsd"|The optional Graphite metric path prefix (after client name)|
|add_path_prefix|boolean|true|If the path_prefix should be used|
|delete_gauges|boolean|false|If gauges that have not been updated should be deleted instead of flushed|
|delete_counters|boolean|false|If counters with a value of 0 should be deleted instead of flushed|
|delete_timers|boolean|false|If timers with a count of 0 should be deleted instead of flushed|
|reset_gauges|boolean|false|If gauges should be reset to 0 after flushing|
|reset_counters|boolean|true|If counters should be reset to 0 after flushing|
|reset_timers|boolean|true|If timers should be reset/cleared after flushing|
|handler|string|"graphite"|Handler to use for the Graphite metrics|
|truncate_output|boolean|true|If output should be truncated|

## Example

Test the StatsD TCP socket:

``` shell
echo "orders:1|c" | nc 127.0.0.1 8125
```


CHANGELOG.md
# Change Log

## [Unreleased]
### Added
- The definition of statsd.rb to include additional check attribute from configuration.

## [2.0.0] - 2018-02-01
### Changed
- The definition of statsd.rb to include truncate_output as a configurable option.
(@TQaztec)

### Security
- update 'rubocop' dependency to '~> 0.51.0' per:
https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-8418.
(@TQaztec)

### Breaking Change
- Dropping ruby 2.0 support after updating rubocop dependency.

## [v1.0.0](https://github.com/sensu-extensions/sensu-extensions-statsd/tree/v1.0.0) (2017-03-21)
[Full Changelog](https://github.com/sensu-extensions/sensu-extensions-statsd/compare/v0.0.1...v1.0.0)

**Merged pull requests:**

- Adhere to Etsy's specifications and behaviour [\#4](https://github.com/sensu-extensions/sensu-extensions-statsd/pull/4) ([portertech](https://github.com/portertech))
- Fix gauge implementation [\#3](https://github.com/sensu-extensions/sensu-extensions-statsd/pull/3) ([corro](https://github.com/corro))

## [v0.0.1](https://github.com/sensu-extensions/sensu-extensions-statsd/tree/v0.0.1) (2016-12-09)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


spec/helpers.rb
require 'rspec'
require 'eventmachine'
require 'sensu/logger'
require 'sensu/settings'
require 'securerandom'

module Helpers
  def timer(delay)
    periodic_timer = EM::PeriodicTimer.new(delay) do
      yield
      periodic_timer.cancel
    end
  end

  def async_wrapper
    EM.run do
      timer(10) do
        raise 'test timed out'
      end
      yield
    end
  end

  def async_done
    EM.stop_event_loop
  end

  def logger
    Sensu::Logger.get(log_level: :fatal)
  end

  def settings
    Sensu::Settings.get
  end

  def epoch
    Time.now.to_i
  end

  def client_template
    {
      name: 'i-424242',
      address: '127.0.0.1',
      subscriptions: [
        'test'
      ],
      timestamp: epoch
    }
  end

  def check_template # rubocop:disable Metrics/MethodLength
    {
      name: 'test',
      type: 'standard',
      issued: epoch,
      command: 'echo WARNING && exit 1',
      subscribers: [
        'test'
      ],
      interval: 1,
      output: 'WARNING',
      status: 1,
      executed: epoch,
      history: [0, 1]
    }
  end

  # TODO: come back and refactor me
  def event_template # rubocop:disable Metrics/MethodLength
    one_second_ago = epoch - 1
    {
      id: ::SecureRandom.uuid,
      client: client_template,
      check: check_template,
      last_ok: one_second_ago,
      last_state_change: one_second_ago,
      occurrences: 1,
      occurrences_watermark: 1,
      action: :create
    }
  end
end


spec/statsd_spec.rb
require File.join(File.dirname(__FILE__), 'helpers')
require 'sensu/extensions/statsd'

describe 'Sensu::Extension::StatsD' do
  include Helpers

  before do
    @extension = Sensu::Extension::StatsD.new
    @extension.settings = {
      client: {
        name: 'foo'
      },
      statsd: {
        flush_interval: 1
      }
    }
    @extension.logger = Sensu::Logger.get(log_level: :fatal)
  end

  it 'can run' do
    async_wrapper do
      @extension.safe_run do |output, status|
        expect(output).to eq('')
        expect(status).to eq(0)
        async_done
      end
    end
  end

  it 'can create graphite plaintext metrics' do
    async_wrapper do
      timer(1) do
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|g'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.open_datagram_socket('127.0.0.1', 0, nil) do |socket|
          data = 'udp:2|g'
          socket.send_datagram(data, '127.0.0.1', 8125)
          socket.close_connection_after_writing
        end
        timer(3) do
          @extension.safe_run do |output, status|
            expect(output).to match(/foo\.statsd\.gauges\.tcp 1\.0/)
            expect(output).to match(/foo\.statsd\.gauges\.udp 2\.0/)
            expect(status).to eq(0)
            async_done
          end
        end
      end
    end
  end

  it 'can support relative gauges' do
    async_wrapper do
      timer(1) do
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:+4|g'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.open_datagram_socket('127.0.0.1', 0, nil) do |socket|
          data = 'udp:-2|g'
          socket.send_datagram(data, '127.0.0.1', 8125)
          socket.close_connection_after_writing
        end
        EM.open_datagram_socket('127.0.0.1', 0, nil) do |socket|
          data = 'udp:-2|g'
          socket.send_datagram(data, '127.0.0.1', 8125)
          socket.close_connection_after_writing
        end
        timer(3) do
          @extension.safe_run do |output, status|
            expect(output).to match(/foo\.statsd\.gauges\.tcp 4\.0/)
            expect(output).to match(/foo\.statsd\.gauges\.udp -4\.0/)
            expect(status).to eq(0)
            async_done
          end
        end
      end
    end
  end

  it 'can support counters with sampling' do
    async_wrapper do
      timer(1) do
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|c|@0.1'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.open_datagram_socket('127.0.0.1', 0, nil) do |socket|
          data = 'udp:1|c'
          socket.send_datagram(data, '127.0.0.1', 8125)
          socket.close_connection_after_writing
        end
        EM.open_datagram_socket('127.0.0.1', 0, nil) do |socket|
          data = 'udp:1|c'
          socket.send_datagram(data, '127.0.0.1', 8125)
          socket.close_connection_after_writing
        end
        timer(3) do
          @extension.safe_run do |output, status|
            expect(output).to match(/foo\.statsd\.counters\.tcp 10/)
            expect(output).to match(/foo\.statsd\.counters\.udp 2/)
            expect(status).to eq(0)
            async_done
          end
        end
      end
    end
  end

  it 'can support timers with sampling' do
    async_wrapper do
      timer(1) do
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:320|ms|@0.1'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.open_datagram_socket('127.0.0.1', 0, nil) do |socket|
          data = 'udp:360|ms'
          socket.send_datagram(data, '127.0.0.1', 8125)
          socket.close_connection_after_writing
        end
        EM.open_datagram_socket('127.0.0.1', 0, nil) do |socket|
          data = 'udp:385|ms'
          socket.send_datagram(data, '127.0.0.1', 8125)
          socket.close_connection_after_writing
        end
        timer(3) do
          @extension.safe_run do |output, status|
            expect(output).to match(/foo\.statsd\.timers\.tcp.lower 3200\.0/)
            expect(output).to match(/foo\.statsd\.timers\.tcp.upper 3200\.0/)
            expect(output).to match(/foo\.statsd\.timers\.tcp.upper_90 3200\.0/)
            expect(output).to match(/foo\.statsd\.timers\.tcp.mean 3200\.0/)
            expect(output).to match(/foo\.statsd\.timers\.udp.lower 360\.0/)
            expect(output).to match(/foo\.statsd\.timers\.udp.upper 385\.0/)
            expect(output).to match(/foo\.statsd\.timers\.udp.upper_90 385\.0/)
            expect(output).to match(/foo\.statsd\.timers\.udp.mean 372\.5/)
            expect(status).to eq(0)
            async_done
          end
        end
      end
    end
  end

  it "can behave like etsy's implementation (i.e resets) with defaults" do
    async_wrapper do
      timer(1) do
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|g'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|c'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|ms'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        timer(3) do
          @extension.safe_run do |output, status|
            expect(output.split("\n").size).to be > 7
            expect(status).to eq(0)
            async_done
          end
        end
      end
    end
  end

  it 'can support deleting gauges, counters, and timers on flush' do
    @extension.settings = {
      client: {
        name: 'foo'
      },
      statsd: {
        flush_interval: 1,
        delete_gauges: true,
        delete_counters: true,
        delete_timers: true,
        truncate_output: false
      }
    }
    async_wrapper do
      timer(1) do
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|g'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|c'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        EM.connect('127.0.0.1', 8125, nil) do |socket|
          data = 'tcp:1|ms'
          socket.send_data(data)
          socket.close_connection_after_writing
        end
        timer(3) do
          @extension.safe_run do |output, status|
            expect(output.split("\n").size).to eq(7)
            expect(status).to eq(0)
            async_done
          end
        end
      end
    end
  end

  it 'can include custom check attributes' do
    @extension.settings = {
      client: {
        name: 'foo'
      },
      statsd: {
        additional_attributes: { attr1: 'value' },
      }
    }
    expect(@extension.definition).to include({ attr1: 'value' })
  end

end


spec/template_spec.rb
require File.join(File.dirname(__FILE__), "helpers")
require "sensu/extensions/template"

describe "Sensu::Extension::Template" do
  include Helpers

  before do
    @extension = Sensu::Extension::Template.new
  end

  it "can run" do
    @extension.safe_run(event_template) do |output, status|
      expect(output).to eq("template")
      expect(status).to eq(0)
    end
  end
end


lib/sensu/extensions/statsd.rb
require 'sensu/extension'

module Sensu
  module Extension
    class StatsDSimpleSocket < EM::Connection
      attr_accessor :data

      def receive_data(data)
        @data << data
      end
    end

    class StatsD < Check # rubocop:disable Metrics/ClassLength
      def name
        'statsd'
      end

      def description
        'a statsd implementation'
      end

      def options # rubocop:disable Metrics/MethodLength
        return @options if @options
        @options = {
          bind: '127.0.0.1',
          port: 8125,
          flush_interval: 10,
          send_interval: 30,
          percentile: 90,
          delete_gauges: false,
          delete_counters: false,
          delete_timers: false,
          reset_gauges: false,
          reset_counters: true,
          reset_timers: true,
          add_client_prefix: true,
          path_prefix: 'statsd',
          add_path_prefix: true,
          handler: 'graphite',
          truncate_output: true
        }
        @options.merge!(@settings[:statsd]) if @settings[:statsd].is_a?(Hash)
        @options
      end

      def definition
        check_attributes = {
          type: 'metric',
          name: name,
          interval: options[:send_interval],
          standalone: true,
          output_format: 'graphite_plaintext',
          handler: options[:handler],
          truncate_output: options[:truncate_output]
        }
        options[:additional_attributes] ? check_attributes.merge!(options[:additional_attributes]) : check_attributes
      end

      def post_init
        @flush_timers = []
        @data = EM::Queue.new
        @gauges = Hash.new { |h, k| h[k] = 0 }
        @counters = Hash.new { |h, k| h[k] = 0 }
        @timers = Hash.new { |h, k| h[k] = [] }
        @metrics = []
        setup_flush_timers
        setup_parser
        setup_statsd_socket
      end

      def run
        output = ''
        if @metrics
          output << @metrics.join("\n") + "\n" unless @metrics.empty?
          @logger.info('statsd collected metrics', count: @metrics.count)
          @metrics = []
        end
        yield output, 0
      end

      private

      def clean(hash, delete = false, reset = false)
        if delete
          hash.delete_if do |_key, value|
            value == 0 || value == []
          end
        end
        if reset
          hash.each do |key, value|
            hash[key] = (value.is_a?(Array) ? [] : 0)
          end
        end
      end

      def add_metric(*args) # rubocop:disable Metrics/MethodLength
        value = args.pop
        path = []
        path << @settings[:client][:name] if options[:add_client_prefix]
        path << options[:path_prefix] if options[:add_path_prefix]
        path = (path + args).join('.')
        if path !~ /^[A-Za-z0-9\._-]*$/
          @logger.info('invalid statsd metric', reason: 'metric path must only consist of alpha-numeric characters, periods, underscores, and dashes',
                                                path: path,
                                                value: value)
        else
          @logger.debug('adding statsd metric', path: path,
                                                value: value)
          @metrics << [path, value, Time.now.to_i].join(' ')
        end
      end

      def flush!
        @gauges.each do |name, value|
          unless value == 0 && options[:delete_gauges]
            add_metric('gauges', name, value)
          end
        end
        clean(@gauges, options[:delete_gauges], options[:reset_gauges])
        @counters.each do |name, value|
          unless value == 0 && options[:delete_counters]
            add_metric('counters', name, value.to_i)
          end
        end
        clean(@counters, options[:delete_counters], options[:reset_counters])
        @timers.each do |name, values|
          next if values.empty? && options[:delete_timers]
          values.sort!
          length = values.length
          min = values.first || 0
          max = values.last || 0
          mean = min
          max_at_threshold = min
          percentile = options[:percentile]
          if length > 1
            threshold_index = ((100 - percentile) / 100.0) * length
            threshold_count = length - threshold_index.round
            valid_values = values.slice(0, threshold_count)
            max_at_threshold = valid_values[-1]
            sum = 0
            valid_values.each { |v| sum += v }
            mean = sum / valid_values.length
          end
          add_metric('timers', name, 'lower', min)
          add_metric('timers', name, 'mean', mean)
          add_metric('timers', name, 'upper', max)
          add_metric('timers', name, "upper_#{percentile}", max_at_threshold)
        end
        clean(@timers, options[:delete_timers], options[:reset_timers])
        @logger.debug('flushed statsd metrics')
      end

      def setup_flush_timers
        @flush_timers << EM::PeriodicTimer.new(options[:flush_interval]) do
          flush!
        end
      end

      # TODO: come back and refactor me
      def setup_parser # rubocop:disable Metrics/MethodLength
        parser = proc do |data|
          begin
            nv, type, raw_sample = data.strip.split('|')
            name, raw_value = nv.split(':')
            value = Float(raw_value)
            sample = Float(raw_sample ? raw_sample.split('@').last : 1)
            case type
            when 'g'
              if raw_value.start_with?('+')
                @gauges[name] += value
              elsif raw_value.start_with?('-')
                @gauges[name] -= value.abs
              else
                @gauges[name] = value
              end
            when /^c/, 'm'
              @counters[name] += value * (1 / sample)
            when 'ms', 'h', 't'
              @timers[name] << value * (1 / sample)
            end
          rescue => error
            @logger.error('statsd parser error', error: error.to_s)
          end
          EM.next_tick do
            @data.pop(&parser)
          end
        end
        @data.pop(&parser)
      end

      def setup_statsd_socket
        @logger.debug('binding statsd tcp and udp sockets', options: options)
        bind = options[:bind]
        port = options[:port]
        EM.start_server(bind, port, StatsDSimpleSocket) do |socket|
          socket.data = @data
        end
        EM.open_datagram_socket(bind, port, StatsDSimpleSocket) do |socket|
          socket.data = @data
        end
      end
    end
  end
end
